### PR TITLE
Make SIA-R87 automatically pass links that target descendants at the very beginning of a `main` landmark

### LINKED ISSUE #706: SIA-R87 should automatically pass links that target descendants at the very beginning of a `main` landmark
SIA-R87 currently requires external input to determine if the following case passes:

```html
<a href="#content">Skip to content</a>

<main>
  <div id="content">
    <p>This is the content</p>
  </div>
</main>
```

Another variation is this:

```html
<a href="#content">Skip to content</a>

<main>
  <div id="content"></div>
  <p>This is the content</p>
</main>
```

It also isn't unthinkable that we might encounter cases like this:

```html
<a href="#content">Skip to content</a>

<div id="content">
  <main>
    <p>This is the content</p>
  </main>
</div>
```

Or even this:

```html
<a href="#content">Skip to content</a>

<div id="content"></div>
<main>
  <p>This is the content</p>
</main>
```

- [ ] Follow up on DEV-11237.

### PR DESCRIPTION
**Pull request by Jym77**
Resolves #706

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,10 +29,16 @@ Items that are related, such as breaking changes, new features, or changes to ex
 
 - [@siteimprove/alfa-selective](packages/alfa-selective): `Selective.exhaust()` is now available in addition to an implementation of `Monad<T>` for `Selective<S, T>`.
 
+- [@siteimprove/alfa-rules](packages/alfa-rules): A predicate `atTheStart` is now available, checking whether a node it "at the start" of another (no actual content between them). ([#735](../../pull/735))
+
```
Comment (kasperisager) [inline CHANGELOG.md:33]:
```suggestion
```

This is an internal change.

File: CHANGELOG.md [hunk 2]
```diff
@@ -29,10 +29,16 @@ Items that are related, such as breaking changes, new features, or changes to ex
...
 ### Changed
 
 - [@siteimprove/alfa-aria](packages/alfa-aria): We now track the latest Editor's Draft of WAI-ARIA, which contains several fixes already implemented by browsers and assistive technologies. ([#728](../../pull/728))
 
+### Fixed
+
+- [@siteimprove/alfa-rules](packages/alfa-rules): R87 now automatically passes if a skip link references an element "at the start" of an element with a role of `main`. ([#735](../../pull/735))
```
Comment (kasperisager) [inline CHANGELOG.md:40]:
```suggestion
- [@siteimprove/alfa-rules](packages/alfa-rules): SIA-R87 now automatically passes if a skip link references an element "at the start" of an element with a role of `main`. ([#735](../../pull/735))
```

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: packages/alfa-rules/src/common/predicate/is-at-the-start.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+import { Refinement } from "@siteimprove/alfa-refinement";
+
+import { isPerceivable } from "./is-perceivable";
+import { isReplaced } from "./is-replaced";
+
+const { equals, or, test } = Predicate;
+const { and } = Refinement;
+
+/**
+ * @see https://act-rules.github.io/glossary/#just-before
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-at-the-start.ts:14]:
```suggestion
 * {@link https://act-rules.github.io/glossary/#just-before}
```

File: packages/alfa-rules/src/common/predicate/is-at-the-start.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ * * [N2, …, N1] => N1 is at the start of N2 if there is no perceivable in [N2, …, N1[
+ * In both cases, jumping to N1 give access to all the perceivable nodes after N2, and no more.
+ *
+ * Perceivable containers (e.g. <div><span>Hello</span></div>) are actually OK, only perceivable actual content is bad.
+ *
+ * Note that the def is actually symmetrical, but N2 is conceptually a container (`<main>`)
+ * while N1 is a single point (an anchor), hence the asymmetry in the code.
+ *
+ * Complexity: the size of the subtree anchored at the lowest common ancestor.
+ **/
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-at-the-start.ts:34]: Make sure to break comments at 80 characters. This applies to the rest of the comments as well.

File: packages/alfa-rules/src/sia-r87/rule.ts [hunk 5]
```diff
@@ -61,6 +64,11 @@ export default Rule.Atomic.of<Page, Document, Question>({
             )
           );
 
+        // there can be more than one element with a role of main, going to any of these is OK.
+        const mains = document
+          .inclusiveDescendants({ flattened: true, nested: true })
```
Comment (kasperisager) [inline packages/alfa-rules/src/sia-r87/rule.ts:69]:
```suggestion
          .inclusiveDescendants({ flattened: true })
```

References cannot cross document boundaries.

File: packages/alfa-rules/src/common/predicate/is-at-the-start.ts (hunk 4 above)
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-at-the-start.ts:34]:
```suggestion
 */
```

**Review by kasperisager — DISMISSED** [review]

File: packages/alfa-rules/src/common/predicate/is-at-the-start.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    next2 = ancestors2.first();
+
+    if (next1.isNone() || next2.isNone()) {
+      // This is needed for the corner case of nodes in different trees but
+      // at the exact same depth…
+      break;
+    }
+  }
+
+  return commonAncestor;
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-at-the-start.ts:118]: I've added `Indexed#reduceWhile()` and `Indexed#reduceUntil()` which should help make this bit of code a little more clear.

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: packages/alfa-rules/src/common/predicate/is-at-the-start.ts (hunk 6 above)
Comment (Jym77) [reply packages/alfa-rules/src/common/predicate/is-at-the-start.ts:118]: The problem is that we need to move in both lists simultaneously.
If we use a reducer on one, we still need to manually avance through the second and I'm not convinced that breaking the symmetry will help legibility 🤔 

OTOH, the last `if` can be replaced by `next1.isSome()` or `not(ancestors1.isEmpty())` in the `while` condition that may make the intend clearer ("while the lists are not empty and start with the same element"). Also maybe renaming `next1/2` to `first1/2`?

Comment (kasperisager) [reply packages/alfa-rules/src/common/predicate/is-at-the-start.ts:118]: I forgot to mention `Indexed#zip()`; first, zip up the ancestor lists and reduce them to the next ancestor while they're equal.

File: packages/alfa-rules/src/common/predicate/is-at-the-start.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+): Option<Node> {
+  return node1
+    .inclusiveAncestors(options)
+    .reverse()
+    .zip(node2.inclusiveAncestors(options).reverse())
+    .reduceWhile<Option<Node>>(
+      ([first1, first2]) => first1.equals(first2),
+      (_, [node]) => Option.of(node),
+      None
+    );
```
Comment (Jym77) [inline packages/alfa-rules/src/common/predicate/is-at-the-start.ts:104]: Actually,
```suggestion
    .takeWhile(([first1, first2]) => first1.equals(first2))
    .last()
    .map(([node,]) =>node)
```
is even more readable, but it does go through the sequence one more time to compute `last()` 😕

**Review by kasperisager — DISMISSED** [review]

**Review by kasperisager — APPROVED** [review]

**Pull request merged**